README.md
# Healthsearch

Welcome to the Healthsearch Demo, an open-source project aimed at showcasing the potential of leveraging user-written reviews and queries to retrieve supplement products based on specific health effects. 

[![Weaviate](https://img.shields.io/static/v1?label=%E2%9D%A4%20made%20with&message=Weaviate&color=green&style=flat-square)](https://weaviate.io/) [![Docker support](https://img.shields.io/badge/Docker_support-%E2%9C%93-4c1?style=flat-square&logo=docker&logoColor=white)](https://docs.docker.com/get-started/) [![Demo](https://img.shields.io/badge/Check%20out%20the%20demo!-blue?&style=flat-square&logo=react&logoColor=white)](https://healthsearch-frontend.onrender.com/) [![Weaviate](https://img.shields.io/static/v1?label=version&message=v1.1&color=pink&style=flat-square)](https://weaviate.io/)

![Demo of Health Search](https://github.com/weaviate/healthsearch-demo/blob/main/frontend/public/Intro.gif)

## 🎯 Overview

The search functionality in this demo accepts natural language queries that are translated into GraphQL queries using LLMs. These GraphQL queries are then utilized to retrieve supplements from a Weaviate database. The demo also exhibits an example of generative search by providing product summaries generated based on the retrieved objects.

> ⚠️ **Disclaimer**: Healthsearch is a technical demonstration, and the results shown should not be treated as health advice. The results and generated summaries are purely based on user-written reviews.

### 💡 Natural Language Translation to GraphQL

We use Large Language Models (LLM), like GPT4, to translate natural language queries into a structured query format, called a GraphQL query.
The demo extracts information about filters, sorting, and limits directly from the context of the query. Whether the query is `the top 10 products for glowing skin`, `products for sleep from a specific brand`, or `best-rated products for toothache`, the demo can interpret these queries and generate an appropriate GraphQL query in return.

### 🔎 Semantic Search

Healthsearch relies on the power of semantic search in user reviews. When seeking products that are `good for joint pain`, for instance, Healthsearch scans user reviews for discussions on products that have alleviated joint pain or similar conditions. The results are then aggregated and grouped according to their respective products.

### 💥 Generative Search

After the translation of the query to GraphQL and the retrieval of the most semantically relevant product, we enhance our demo with a feature called `Generative Search`. Essentially, we examine the top five results and employ an LLM to generate a product summary. This concise summary offers a brief overview of the products, highlighting their pros and cons and providing valuable insights. Each summary is crafted around the query, ensuring every search is unique and interesting.

### 🔥 Semantic Cache

We embed the generated results and queries to Weaviate, and use it as a `Semantic Cache`.
This method is advantageous as it enables the demo to return results from queries that are semantically equal to the new query. For example `good for joint pain` and `helpful for joint pain` are semantically very similar and should return the same results, whereas `bad for joint pain` should have its own generated result. This method allows us to gain much more from generated results than traditional string matching would permit. It's a simple yet potent solution that enhances the efficiency of the search process.

## 🔧 Template

This repository is designed to serve as a template - a starting point for your own projects with Weaviate. Take inspiration from how we've implemented certain features and feel free to enhance it in your own project. We welcome comments, ideas, and feedback. Embrace the open-source spirit!

## 💰 Language Learning Model (LLM) Costs

This demonstration primarily uses OpenAI models for embedding supplement products, processing user queries, and generating summaries. By default, any costs associated with using these services will be billed to the access key that you provide.

If you prefer, you can replace the OpenAI models with any other Language Learning Model (LLM) provider. However, please be aware that completely changing the API will require further adjustments to the code.

Below, we provide a rough estimate of the costs involved in importing data to Weaviate. For a comprehensive understanding, please visit OpenAI's pricing page at https://openai.com/pricing.

### Data Embedding Costs
We employ the Ada v2 model for embedding data into the Weaviate cluster. At the time of writing this README, the model costs $0.0001 for every 1k tokens (note that approximately [4 characters equal 1 token](https://help.openai.com/en/articles/4936856-what-are-tokens-and-how-to-count-them)). As a rough approximation, importing the dataset to Weaviate might cost around $0.002.

### Query Construction and Summary Generation Costs
We use the GPT-4 model for building GraphQL queries and generating summaries. As of the time of writing this README, this model costs $0.03/1k tokens for input and $0.06/1k tokens for output. The exact costs are dependent on the user query and the results returned by the GraphQL query. Please take these factors into account when calculating your expected costs. You can also change the `model_name` variable to `gpt-3.5-turbo` inside the `api.py` script in the backend folder. The GPT-3 model costs $0.0015/1k tokens for input and $0.002/1k tokens for output.

## 🛠️ Project Structure

The Healthsearch Demo is structured in three main components:

1. A Weaviate database (either cluster hosted on WCS or local).
2. A FastAPI endpoint facilitating communication between the LLM provider and database.
3. An interactive React frontend for displaying the information.

Make sure you have Python (`>=3.8.0`) and Node (`>=18.16.0`) installed. We also support Docker and provide Dockerfiles for the setup.

## 🐳 Quickstart with Docker

You can use Docker to setup the demo in one line of code! If you're not familiar with Docker you can read more about it here (https://docker-curriculum.com/)

1. **Set environment variables:**
- The following environment variables need to be set
- ```OPENAI_API_KEY=your-openai-api-key```
> Use the `.env` file inside the backend folder to set the variable (https://github.com/theskumar/python-dotenv)
> Note that if you're using the GPT-4 model (by default), ensure your OpenAI key has access. You can change the `model_name` variable to `gpt-3.5-turbo` inside the `api.py` script.

1. **Use docker compose**
-  `docker-compose up`

2. **Access the frontend on:**
- `localhost:3000`


## 📚 Getting Started

To kick-start with the Healthsearch Demo, please refer to the READMEs in the `Frontend` and `Backend` folders:

- [Frontend README](./frontend/README.md)
- [Backend README](./backend/README.md)

## 💡 Usage

Follow these steps to use the Healthsearch Demo:

1. Set up the Weaviate database, FastAPI backend, and the React frontend by following the instructions in their respective READMEs.
2. Launch the database, backend server, and the frontend application.
3. Use the interactive frontend to input your natural language query related to a health condition or benefit.
4. The frontend sends the query to the backend, which transforms the natural language query into a GraphQL query using the LLM.
5. The backend sends the GraphQL query to the Weaviate database to fetch relevant reviews based on the user query.
6. The frontend displays the results, allowing you to explore the most semantic-related supplements to your specific health-related query.

## 💖 Open Source Contribution

Your contributions are always welcome! Feel free to contribute ideas, feedback, or create issues and bug reports if you find any! Please adhere to the code guidelines that include formatting, linting, and testing.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## [1.1.0] - 13.07.2023

### Added
- [x] Display saved cached queries to select from
- [x] Return whether results was cached (direct match, similarity match) or generated

### Fixed
- [x] Adjusted threshhold for semantic cache
- [x] Improved product schema for weaviate 
- [x] Hard filtering based on review count to reduce noise

## [1.0.0] - 03.07.2023

### Added 
- Improved documentation
- Preparation for publish and deployment
- Docker support
- Add more info on frontend about request and cache count

### Fixed

- Responsive Design

## [0.5.0] - 27.06.2023

### Added 
- Implement Weaviate Caching
- Improve robustness and error handling
- Add more products

### Fixed
- Fixed annotation whitespace
- Adjusted schema to exclude product names from vectorizer

## [0.4.0] - 04.06.2023

### Added 
- Generative Search card + workflow

## [0.3.0] - 25.05.2023

### Added 
- Memes
- Added Rive Animations
- Improved and documented API code
- Improved general UI and UX

## [0.2.0] - 23.05.2023

### Added 
- Improved design, background, cleaner look
- Dataset upgrad with pictures, generated summary, description and cleaned ingredientes
- Display detailed product view
- display generated summary
- reviews with condition annotation

## [0.1.0] - 17.05.2023

### Added 
- Add distance label
- Fix refreshing products bug
- Added health disclaimer
- Improved class schema
- Improved prompt length and changed model to gpt3-turbo

## [0.0.3] - 16.05.2023

### Added 
 - Better Product Viz
 - Fix loading circle position
 - Create Cache when not detected
 - Better Title Design
 - Add API Status
 - Darkmode Scrollbars
 - Changed background

## [0.0.2] - 15.05.2023

### Added 
 - Added pictures to the dataset
 - Infoboxes for the left cards
 - Cache results in API
 - Data Preparation in API (add more robustness)
 - Basic Product visualization


## [0.0.1] - 11.05.2023

### Added 

- Implemented query generation over API
- Present search results
- Copy clipboard functionality
- Create readme's

## [0.0.0] - 10.05.2023

### Added 

- Started project
- Added Healthsea dataset
- Imported dataset to Weaviate
- Started frontend
- Added Natural Language Input
- Added Query Output with Syntax Highlighting
- Added basic demo layout

backend/clear_cache.py
import weaviate  # type: ignore[import]
import typer
import os

from pathlib import Path
from wasabi import msg  # type: ignore[import]

from dotenv import load_dotenv

load_dotenv()


def main() -> None:
    msg.divider("Starting clearing cache")

    # Connect to Weaviate
    url = os.environ.get("HEALTHSEARCH_SERVER", "")
    openai_key = os.environ.get("OPENAI_API_KEY", "")
    auth_config = weaviate.AuthApiKey(
        api_key=os.environ.get("HEALTHSEARCH_API_KEY", "")
    )

    if openai_key == "" or url == "":
        msg.fail("Environment Variables not set.")
        msg.info(f"URL: {url}")
        msg.info(f"OPENAI API KEY: {openai_key}")
        return

    client = weaviate.Client(
        url=url,
        additional_headers={"X-OpenAI-Api-Key": openai_key},
        auth_client_secret=auth_config,
    )

    msg.good("Client connected to Weaviate Instance")

    cache_obj = {
        "class": "CachedResult",
        "description": "Cached results",
        "properties": [
            {
                "dataType": ["text"],
                "description": "GraphQL Query",
                "name": "graphQuery",
                "moduleConfig": {
                    "text2vec-openai": {
                        "skip": True,
                        "vectorizePropertyName": False,
                    }
                },
            },
            {
                "dataType": ["text"],
                "description": "Natural Language Query",
                "name": "naturalQuery",
                "moduleConfig": {
                    "text2vec-openai": {
                        "skip": False,
                        "vectorizePropertyName": True,
                    }
                },
            },
            {
                "dataType": ["text"],
                "description": "Retrieved Products",
                "name": "products",
                "moduleConfig": {
                    "text2vec-openai": {
                        "skip": True,
                        "vectorizePropertyName": False,
                    }
                },
            },
            {
                "dataType": ["text"],
                "description": "Generated Summary",
                "name": "summary",
                "moduleConfig": {
                    "text2vec-openai": {
                        "skip": True,
                        "vectorizePropertyName": False,
                    }
                },
            },
        ],
        "vectorizer": "text2vec-openai",
    }

    if not client.schema.exists("CachedResult"):
        client.schema.create_class(cache_obj)
    else:
        client.schema.delete_class("CachedResult")
        client.schema.create_class(cache_obj)

    msg.good("Cache cleared")


if __name__ == "__main__":
    typer.run(main)


backend/api.py
import openai
import os
import weaviate  # type: ignore[import]
import json
import re

from wasabi import msg  # type: ignore[import]

from fastapi import FastAPI, status
from fastapi.responses import JSONResponse
from fastapi.middleware.cors import CORSMiddleware
from pydantic import BaseModel

from dotenv import load_dotenv

load_dotenv()

# Request Count
request_count = 0
cache_count = 0

# Configuration
data_fields = [
    "name",
    "brand",
    "ingredients",
    "reviews",
    "image",
    "rating",
    "description",
    "summary",
    "effects",
]

model_name = (
    "gpt-4"  # default (gpt-4), change to (gpt-3.5-turbo) if you don't have access
)

# Define OpenAI API key, Weaviate URL, and auth configuration
openai.api_key = os.environ.get("OPENAI_API_KEY", "")
url = os.environ.get("HEALTHSEARCH_SERVER", "")
auth_config = weaviate.AuthApiKey(api_key=os.environ.get("HEALTHSEARCH_API_KEY", ""))

# Setup OpenAI API key and Weaviate client
if openai.api_key != "":
    msg.good("Open AI API Key available")
    if url:
        client = weaviate.Client(
            url=url,
            additional_headers={"X-OpenAI-Api-Key": openai.api_key},
            auth_client_secret=auth_config,
        )
    else:
        msg.warn("Server URL not available")
        exit()
else:
    msg.warn("Open AI API Key not available")
    exit()

# Define system prompt for conversation with GPT model
system_prompt = """

You are a parser that understands the meaning of natural language queries and parses them into valid graphql queries based on this schema:

    class_obj = {
        "class": "Product",
        "description": "Supplementary products from iHerb",
        "properties": [
            {
                "dataType": ["text"],
                "description": "The name of the product",
                "name": "name",
            },
            {
                "dataType": ["text"],
                "description": "The brand of the product",
                "name": "brand",
            },
            {
                "dataType": ["text"],
                "description": "The ingredients contained in the product.",
                "name": "ingredients",
            },
            {
                "dataType": ["text[]"],
                "description": "Reviews about the product",
                "name": "reviews",
            },
            {
                "dataType": ["text"],
                "description": "Image URL of the product",
                "name": "image",
            },
            {
                "dataType": ["number"],
                "description": "The Rating of the product",
                "name": "rating",
            },
            {
                "dataType": ["text"],
                "description": "The description of the product",
                "name": "description",
            },
            {
                "dataType": ["text"],
                "description": "The summary of the reviews",
                "name": "summary",
            },
            {
                "dataType": ["text"],
                "description": "The health effects of the product",
                "name": "effects",
            },
        ],
    }

    The query will be used to retrieve supplement products from a Weaviate database, make sure that all fields are returned with the _additional distance attribute.
    Your answers are only allowed to contain the query, the results will be used directly.

    Example natural language query: 'Which product is helpful for joint pain?' produce this GraphQL query:

    {
      Get {
        Product(
          nearText: {concepts: ["Helpful", "joint pain"]}
        ) {
          name
          brand
          ingredients
          reviews
          image
          rating
          description
          summary
          effects
          _additional {
            id
            distance
          }
        }
      }
    }

    Example natural language query: 'Products from brand "Life Extension" for glowing skin' produce this GraphQL query:

    {
  Get {
    Product(
      nearText: {concepts: ["glowing skin"]}
      where: {
        path: ["brand"],
        operator: Equal,
        valueString: "Life Extension"
      }
    ) {
          name
          brand
          ingredients
          reviews
          image
          rating
          description
          summary
          effects
      _additional {
      id
        distance
      }
    }
  }
}

  Example natural language query: 'Lowest rated products for energy' produce this GraphQL query:

  {
  Get {
    Product(
      nearText: {concepts: ["energy"]}
      sort: [{
      path: ["rating"]     
      order: asc          
    }]
    ) {
      name
          brand
          ingredients
          reviews
          image
          rating
          description
          summary
          effects
      _additional {
      id
        distance
      }
    }
  }
}

"""

# FastAPI App
app = FastAPI()

origins = ["http://localhost:3000", "https://healthsearch-frontend.onrender.com"]

# Add middleware for handling Cross Origin Resource Sharing (CORS)
app.add_middleware(
    CORSMiddleware,
    allow_origins=origins,
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
)


def handle_results(results: dict) -> list:
    """Process the results from Weaviate to a defined format
    @parameter results : dict - Dict containing the products retrieved from Weaviate
    @returns list - A list of products in dict format
    """
    try:
        end_results = []
        data = results["data"]
        for key in data:
            query_results = data[key]["Product"]
            for query_result in query_results:
                # Add hard filter
                if len(query_result.get("reviews", [])) >= 5:
                    end_results.append(
                        {
                            "brand": query_result.get("brand", "No brand"),
                            "name": query_result.get("name", "No name"),
                            "rating": query_result.get("rating", 0.0),
                            "ingredients": query_result.get("ingredients", ""),
                            "description": query_result.get("description", ""),
                            "summary": query_result.get("summary", ""),
                            "effects": query_result.get("effects", ""),
                            "reviews": query_result.get("reviews", []),
                            "image": query_result.get("image", ""),
                            "distance": round(
                                query_result.get("_additional", {"distance": 0})[
                                    "distance"
                                ],
                                2,
                            ),
                        }
                    )
        return end_results

    except Exception as e:
        msg.fail("Error at handling results")
        print(e)
        print(results)
        return [
            {
                "brand": "Brand",
                "name": "Product",
                "rating": 0.0,
                "ingredients": "Substances",
                "description": "description",
                "summary": "summary",
                "effects": "effects",
                "reviews": ["Review"],
                "image": "",
                "distance": 0.0,
            }
        ]


def get_cache(natural_query: str) -> dict:
    """Check if a natural language query exists in the Weaviate database
    @parameter natural_query : str - Natural Query from the user
    @returns dict - Data object retrieved from weaviate
    """
    filter = {
        "path": ["naturalQuery"],
        "operator": "Equal",
        "valueText": str(natural_query),
    }

    results = (
        client.query.get(
            "CachedResult", ["naturalQuery", "graphQuery", "products", "summary"]
        )
        .with_where(filter)
        .with_limit(1)
        .do()
    )

    if "errors" in results:
        msg.warn(f"Error in get_cache: {results}")
        return {"data": {"Get": {"CachedResult": []}}}

    if results["data"]["Get"]["CachedResult"]:
        if natural_query == results["data"]["Get"]["CachedResult"][0]["naturalQuery"]:
            return results

    return {"data": {"Get": {"CachedResult": []}}}


def get_cache_count() -> list:
    """Update the global cache count and return all cached queries
    @returns list of queries
    """
    query = client.query.get("CachedResult", ["naturalQuery"]).do()
    cachedQueries = [
        naturalQuery["naturalQuery"]
        for naturalQuery in query["data"]["Get"]["CachedResult"]
    ]
    return cachedQueries


def check_cache(cache_results: dict, natural_query: str, max_distance: float) -> dict:
    """Check if retrieved results are empty and use semantic search to find similar cached results based on the natural query
    @parameter cache_results : dict - Weaviate retrieved results
    @parameter natural_query : str - Natural Query of the user
    @parameter max_distance : float - Distance threshold for semantic search
    @returns dict | None - Data object retrieved from weaviate
    """
    if cache_results["data"]["Get"]["CachedResult"]:
        msg.good("Cache entry exists!")
        cache_results["data"]["Get"]["CachedResult"][0]["summary"] = (
            "🛰️ RETRIEVED FROM CACHE: "
            + cache_results["data"]["Get"]["CachedResult"][0]["summary"]
        )
        return cache_results
    else:
        msg.warn("Cache entry does not exist!")
        nearText = {"concepts": [natural_query], "max_distance": max_distance}
        results = (
            client.query.get(
                "CachedResult", ["naturalQuery", "graphQuery", "products", "summary"]
            )
            .with_near_text(nearText)
            .with_limit(1)
            .with_additional(["distance"])
            .do()
        )
        if not results["data"]["Get"]["CachedResult"]:
            msg.warn("No similar cache entry match")
            return {}
        elif (
            results["data"]["Get"]["CachedResult"][0]["_additional"]["distance"]
            > max_distance
        ):
            msg.warn("No similar cache entry match")
            return {}
        else:
            msg.good(
                f'Retrieved similar results (distance {results["data"]["Get"]["CachedResult"][0]["_additional"]["distance"]})'
            )
            results["data"]["Get"]["CachedResult"][0]["summary"] = (
                f"⭐ RETURNED SIMILAR CACHED RESULTS FROM QUERY '{results['data']['Get']['CachedResult'][0]['naturalQuery']}' ({round(results['data']['Get']['CachedResult'][0]['_additional']['distance'],2)}) : "
                + results["data"]["Get"]["CachedResult"][0]["summary"]
            )
            return results


def add_cache(naturalQuery: str, graphQuery: str, results: dict, summary: str) -> None:
    """Add results to the Weaviate cache
    @parameter natural_query : str - Natural Query of the user
    @parameter graphQuery : str - Generated GraphQL query
    @parameter results : dict - Results retrieved from Weaviate
    @parameter summary : str - Generated product summary
    @returns None
    """
    data_object = {
        "graphQuery": graphQuery,
        "naturalQuery": naturalQuery,
        "products": json.dumps(results),
        "summary": summary,
    }

    with client.batch as batch:
        batch.batch_size = 1
        client.batch.add_data_object(data_object, "CachedResult")

    msg.good("Added new cache entry")


def modify_graphql(graphQuery: str, natural_query: str, fields: list) -> str:
    """Uses the generated GraphQL query to generate the query for creating the product summary
    @parameter natural_query : str - Natural Query of the user
    @parameter graphQuery : str - Generated GraphQL query
    @parameter fields : list - List of strings for every data field
    @returns str - GraphQL query with the generate module
    """
    # Fields to keep
    fields_to_keep = ["summary", "description", "ingredients"]

    # All fields present in the query
    all_fields = fields

    # Fields to remove
    fields_to_remove = [field for field in all_fields if field not in fields_to_keep]

    # Pattern to match 'where' clause
    where_pattern = r"where:\s*\{([^\}]*)\}"

    # Check if 'where' field exists
    where_match = re.search(where_pattern, graphQuery, flags=re.DOTALL)
    if where_match:
        # 'where' field exists, extract field names within 'where' clause
        where_clause = where_match.group(1)
        where_fields = re.findall(r"path:\s*\[([^\]]*)\]", where_clause)

        # If a field in the 'where' clause is in the fields to be removed, keep it
        for where_field in where_fields:
            where_field = where_field.replace('"', "").strip()
            if where_field in fields_to_remove:
                fields_to_remove.remove(where_field)

    # Pattern to match 'sort' clause
    sort_pattern = r"sort:\s*\[\{([^\}]*)\}\]"

    # Check if 'sort' field exists
    sort_match = re.search(sort_pattern, graphQuery, flags=re.DOTALL)
    if sort_match:
        # 'sort' field exists, extract field names within 'sort' clause
        sort_clause = sort_match.group(1)
        sort_fields = re.findall(r"path:\s*\[([^\]]*)\]", sort_clause)

        # If a field in the 'sort' clause is in the fields to be removed, keep it
        for sort_field in sort_fields:
            sort_field = sort_field.replace('"', "").strip()
            if sort_field in fields_to_remove:
                fields_to_remove.remove(sort_field)

    # Remove unwanted fields
    for field in fields_to_remove:
        graphQuery = re.sub(rf"\s*{field}", "", graphQuery)

    # Pattern to match '_additional'
    pattern = r"(_additional\s*\{[^\}]*\})"

    # Pattern to match 'limit'
    limit_pattern = r"(limit:\s*\d+)"

    # Check if 'limit' field exists and if it's greater than 5
    limit_match = re.search(limit_pattern, graphQuery)
    if limit_match:
        # Limit field exists, ensure it's not greater than 5
        limit_value = int(limit_match.group().split(":")[1].strip())
        if limit_value > 5:
            graphQuery = re.sub(limit_pattern, "limit: 5", graphQuery)
    else:
        # 'limit' field doesn't exist, add it
        graphQuery = re.sub(r"Product\(", "Product(\n      limit: 5", graphQuery)

    # Replacement '_additional'
    replacement = f"""
      _additional {{
        generate(
          groupedResult: {{
            task: "Summarize products based on this query: {natural_query}"
          }}
        ) {{
          groupedResult
          error
        }}
        id
        distance
      }}"""

    # Replace '_additional'
    modified_query = re.sub(pattern, replacement, graphQuery, flags=re.DOTALL)

    return modified_query


# Class for the Natural Language Query
class NLQuery(BaseModel):
    text: str


# Define health check endpoint
@app.get("/health")
async def root():
    global cache_count

    try:
        cached_queries = get_cache_count()
        cache_count = len(cached_queries)
        return JSONResponse(
            content={
                "message": "Alive!",
                "requests": request_count,
                "cache_count": cache_count,
                "cache_queries": cached_queries,
            }
        )
    except Exception as e:
        msg.fail(f"Healthcheck failed with {str(e)}")
        return JSONResponse(
            content={
                "message": "Database connection failed!",
                "requests": request_count,
                "cache_count": cache_count,
                "cache_queries": [],
            },
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
        )


# Define endpoint for generating GraphQL query from natural language
@app.post("/generate_query")
async def generate_query(payload: NLQuery):
    """Process the Payload sent by the Frontend, send API request to Open AI API, receive and format the results and send them back to the frontend
    @parameter payload : ProcessTweetsPayload - Payload sent by the frontend containing the prompt, tweets and context tags
    @returns JSONResponse - JSON containing the results
    """
    global request_count
    request_count += 1
    query_text = payload.text.strip().lower()

    start_prompt = f"Convert this natural language to a GraphQL Query and only return the query, it will be directly used: {query_text}"

    # Easter Egg
    if query_text == "easteregg":
        return JSONResponse(
            content={
                "query": "🚀 Congratulations, you rolled the demo!",
                "results": {},
                "generative_summary": "You just got rick-rolled...",
            }
        )

    # Cache Retrieval
    results = check_cache(get_cache(query_text), query_text, 0.14)

    if len(results) > 0:
        products = json.loads(results["data"]["Get"]["CachedResult"][0]["products"])

        return JSONResponse(
            content={
                "query": results["data"]["Get"]["CachedResult"][0]["graphQuery"],
                "results": products,
                "generative_summary": results["data"]["Get"]["CachedResult"][0][
                    "summary"
                ],
            }
        )

    # Production
    else:
        prompt = start_prompt
        error_message = ""
        for i in range(0, 3):
            try:
                response = openai.ChatCompletion.create(
                    model=model_name,
                    messages=[
                        {
                            "role": "system",
                            "content": system_prompt,
                        },
                        {"role": "user", "content": prompt},
                    ],
                )
            except Exception as e:
                msg.fail("API Request Failed")
                msg.info(str(e))
                return JSONResponse(
                    content={
                        "query": f"API request failed...",
                        "results": {},
                        "generative_summary": f"💥 Oh no... API request failed: {str(e)}",
                    }
                )

            for choice in response["choices"]:
                content = str(choice["message"]["content"])
                results = client.query.raw(content)

                if "errors" in results:
                    error_message = str(results["errors"])
                    prompt = f"The provided GraphQL is not valid, see this error: {error_message} please fix this GraphQL query for a Weaviate database: {content}"
                    msg.warn(f"({i}) Query Error detected, retrying...")
                    msg.info(prompt)
                    continue

                results = handle_results(results)  # type: ignore[assignment]

                generative_query = modify_graphql(str(content), query_text, data_fields)
                generative_results = client.query.raw(str(generative_query))

                if "errors" in generative_results:
                    generative_summary = str(generative_results["errors"])
                    msg.warn("Generative Query Failed!")
                    return JSONResponse(
                        content={
                            "query": "".join(
                                [
                                    str(content) + "\n\n",
                                    "# Query with generative module \n\n",
                                    generative_query,
                                ]
                            ),
                            "results": results,
                            "generative_summary": generative_summary,
                        }
                    )

                else:
                    generative_summary = str(
                        generative_results["data"]["Get"]["Product"][0]["_additional"][
                            "generate"
                        ]["groupedResult"]
                    )

                add_cache(
                    query_text,
                    "".join(
                        [
                            str(content) + "\n\n",
                            "# Query with generative module \n\n",
                            generative_query,
                        ]
                    ),
                    results,
                    generative_summary,
                )

                return JSONResponse(
                    content={
                        "query": "".join(
                            [
                                str(content) + "\n\n",
                                "# Query with generative module \n\n",
                                generative_query,
                            ]
                        ),
                        "results": results,
                        "generative_summary": "✨ GENERATED: " + generative_summary,
                    }
                )

        return JSONResponse(
            content={
                "query": f"Not able to construct query...",
                "results": {},
                "generative_summary": f"💥 Oh no... We couldn't create a GraphQL query from your input!",
            }
        )
